module/localisation/BallFilter/src/BallFilter.cpp

```diff
@@ -35,53 +37,107 @@ namespace module::localisation {
+            cfg.ukf.initial.mean.position = config["ukf"]["initial"]["mean"]["position"].as<Expression>();
+            cfg.ukf.initial.mean.velocity = config["ukf"]["initial"]["mean"]["velocity"].as<Expression>();
+
+            // Set out initial covariance
+            cfg.ukf.initial.covariance.position = config["ukf"]["initial"]["covariance"]["position"].as<Expression>();
+            cfg.ukf.initial.covariance.velocity = config["ukf"]["initial"]["covariance"]["velocity"].as<Expression>();
+
+            // Set our initial state with the config means and covariances, flagging the filter to reset it
+            cfg.initial_mean.rBWw = cfg.ukf.initial.mean.position;
+            cfg.initial_mean.vBw  = cfg.ukf.initial.mean.velocity;
+
+            cfg.initial_covariance.rBWw = cfg.ukf.initial.covariance.position;
+            cfg.initial_covariance.vBw  = cfg.ukf.initial.covariance.velocity;
+            ukf.set_state(cfg.initial_mean.getStateVec(), cfg.initial_covariance.asDiagonal());
+
+            last_time_update = NUClear::clock::now();
         });
 
         /* To run whenever a ball has been detected */
-        on<Trigger<VisionBalls>, With<Sensors>>().then([this](const VisionBalls& balls, const Sensors& sensors) {
-            if (balls.balls.size() > 0) {
-
-                // Get the first vision ball measurement and treat it as the closest measurement to our estimate
-                Eigen::Vector3f rBCc =
-                    reciprocalSphericalToCartesian(balls.balls[0].measurements[0].srBCc.cast<float>());
-                float lowest_distance = (rBCc - filtered_rBCc).norm();
-
-                // Loop through all our ball measurements and find the closest measurement to our current estimate
-                for (const auto& ball : balls.balls) {
-                    Eigen::Vector3f temp_rBCc =
-                        reciprocalSphericalToCartesian(ball.measurements[0].srBCc.cast<float>());
-                    if ((temp_rBCc - filtered_rBCc).norm() < lowest_distance) {
-                        lowest_distance = (temp_rBCc - filtered_rBCc).norm();
-                        rBCc            = temp_rBCc;
+        on<Trigger<VisionBalls>, With<Sensors>, With<FieldDescription>>().then(
+            [this](const VisionBalls& balls, const Sensors& sensors, const FieldDescription& fd) {
+                if (balls.balls.size() > 0) {
+                    Eigen::Isometry3f Hcw = Eigen::Isometry3f(balls.Hcw.cast<float>());
+                    Eigen::Isometry3f Hwc = Hcw.inverse();
+                    auto state            = BallModel<float>::StateVec(ukf.get_state());
+
+                    // Get the first ball measurement and treat it as the closest measurement to our current estimate
+                    Eigen::Vector3f rBWw =
+                        Hwc * reciprocalSphericalToCartesian(balls.balls[0].measurements[0].srBCc.cast<float>());
+                    float lowest_distance = (rBWw.head<2>() - state.rBWw).norm();
```

Comment:
Use `squaredNorm()` instead (here, as well as below). It will save you doing square root operations and will have no impact on the logic